internal/pkg/service/buffer/api/receive/quota/quota.go

```diff
@@ -33,57 +45,80 @@ type dependencies interface {
 	StatsCache() *statistics.CacheNode
 }
 
-func New(ctx context.Context, wg *sync.WaitGroup, d dependencies) *Quota {
-	q := &Quota{
-		clock:     d.Clock(),
-		config:    d.APIConfig(),
-		stats:     d.StatsCache(),
-		cacheLock: &sync.RWMutex{},
-		cache:     make(bufferedSizeMap),
+func New(ctx context.Context, wg *sync.WaitGroup, d dependencies) *Checker {
+	c := &Checker{
+		clock:         d.Clock(),
+		config:        d.APIConfig(),
+		stats:         d.StatsCache(),
+		nextLogAtLock: &sync.RWMutex{},
+		nextLogAt:     make(map[key.ReceiverKey]time.Time),
+		cacheLock:     &sync.RWMutex{},
+		cache:         make(bufferedSizeMap),
 	}
 
 	// Periodically invalidates the cache.
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		ticker := q.clock.Ticker(q.config.ReceiverBufferSizeCacheTTL)
+		ticker := c.clock.Ticker(c.config.ReceiverBufferSizeCacheTTL)
 		defer ticker.Stop()
 		for {
 			select {
 			case <-ctx.Done():
 				return
 			case <-ticker.C:
-				q.cacheLock.Lock()
-				q.cache = make(bufferedSizeMap, len(q.cache))
-				q.cacheLock.Unlock()
+				c.ClearCache()
 			}
 		}
 	}()
 
-	return q
+	return c
 }
 
 // Check checks whether the size of records that one receiver can buffer in etcd has not been exceeded.
-func (i *Quota) Check(k key.ReceiverKey) error {
+func (c *Checker) Check(k key.ReceiverKey) error {
 	// Load bufferedBytes from the fast cache.
-	i.cacheLock.RLock()
-	buffered, found := i.cache[k]
-	i.cacheLock.RUnlock()
+	c.cacheLock.RLock()
+	buffered, found := c.cache[k]
+	c.cacheLock.RUnlock()
 
 	// If not found, then calculate statistics from the slower cache.
 	if !found {
-		buffered = i.stats.ReceiverStats(k).Buffered.RecordsSize
-		i.cacheLock.Lock()
-		i.cache[k] = buffered
-		i.cacheLock.Unlock()
+		buffered = c.stats.ReceiverStats(k).Buffered.RecordsSize
+		c.cacheLock.Lock()
+		c.cache[k] = buffered
+		c.cacheLock.Unlock()
 	}
 
-	if limit := i.config.ReceiverBufferSize; buffered > limit {
-		return errors.Errorf(
+	if limit := c.config.ReceiverBufferSize; buffered > limit {
+		return commonErrors.NewInsufficientStorageError(c.shouldLogError(k), errors.Errorf(
 			`no free space in the buffer: receiver "%s" has "%s" buffered for upload, limit is "%s"`,
 			k.ReceiverID, buffered.HumanReadable(), limit.HumanReadable(),
-		)
+		))
 	}
 
 	return nil
 }
+
+func (c *Checker) ClearCache() {
+	c.cacheLock.Lock()
+	c.cache = make(bufferedSizeMap, len(c.cache))
+	c.cacheLock.Unlock()
+}
+
+// shouldLogError method determines if the quota error should be logged.
+func (c *Checker) shouldLogError(k key.ReceiverKey) bool {
+	now := c.clock.Now()
+
+	c.nextLogAtLock.RLock()
+	logTime := c.nextLogAt[k] // first time it returns zero time
```

Comment:
Ak dany `k` v map nie je, tak to vrati prazdny `time.Time{}`, co nam vyhovuje.

Namiesto toho by sa mohla pouzit alternativa zapisu:
```
logTime, found := c.nextLogAt[k]
```